lib/recursive-methods.rb

```diff
@@ -1,49 +1,158 @@
+# @PaulineChane on GitHub, Ada C14 CS Fun
+# frozen_string_literal: true
+
 # Authoring recursive algorithms. Add comments including time and space complexity for each method.
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n) Worst case the function calls itself with every decrease from n to 1 by 1 until n <= 1 for a
+#                  total of n calls.
+# Space complexity: O(n) The function makes a recursive call n times, resulting in a stack of n recursive calls.
 def factorial(n)
-    raise NotImplementedError, "Method not implemented"
+  raise ArgumentError, 'Cannot take factorial of number less than 0' if n.negative?
+
+  n <= 1 ? 1 : n * factorial(n - 1)
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n) The function will not exceed calls longer than the length of the string that it is "chopping"
+#                  the last character off of.
+# Space complexity: O(n^2) A new recursion call is called, at worst, for every removal of the last character of the string.
+#                   However, a new string of n length is created to return for 2n space, while every call of reverse
+#                   will take in a new string of max length s.length-1 for approximately 2n*n = 2n^2 space.
 def reverse(s)
```

Comment:
👍 However both the time/space complexities are O(n^2) because chop makes a new string which is a copy of the original (missing the last character).